### ISSUE TITLE
Invalid call to `MaybeUninit<T>::assume_init()` and uninitialized reads.

### ISSUE THREAD

**Issue by icmccorm** [MaterializeInc/rust-dec#76]
I've run a newer version of the Miri extension mentioned in #74, and it's found a few cases where uninitialized memory is being read. Not all of these constitute undefined behavior. Some can be addressed by changes to the Rust encapsulation in `dec`, but others might need to be fixed within the `C` source in `decnumber`. Here's a long narrative, feel free to skip to the TLDR at the bottom.

The first of these situations occurs in test `decnum_raw_parts` when using `MaybeUninit<T>` in [decimal.rs](https://github.com/MaterializeInc/rust-dec/blob/master/dec/src/decimal.rs)
```
error: Undefined Behavior: constructing invalid value at .value.lsu[6]: encountered uninitialized memory, but expected an integer
    --> /Users/icmccorm/git/rust-dec/dec/src/decimal.rs:1924:13
     |
1924 |             d.assume_init()
     |             ^^^^^^^^^^^^^^^ constructing invalid value at .value.lsu[6]: encountered uninitialized memory, but expected an integer
```
It seems like the LSU array isn't fully initialized. This pattern occurs in `total_cmp` and `partial_cmp` methods, and can be fixed by using `Decimal::<N>::zero()` instead of `MaybeUninit<T>`

However, there are also situations where uninitialized memory is being read within the C implementation. This occurs when running `test_decimal128_decomposition`, and it happens on line 1397 in the implementation of `decFloatAdd` within [decBasic.c](https://github.com/MaterializeInc/rust-dec/blob/master/decnumber-sys/decnumber/decBasic.c).
```
UBFROMUI(acc+COFF+DECPMAX+4, UBTOUI(buf+COFF+DECPMAX+4));
```
It seems like this is just due to reading uninitialized bytes from either `buf` or `acc`, which are never *completely* zero-initialized. I added a setting to have the interpreter zero-initialize all stack memory by default, and when I enable this, all tests that previously failed at this line will pass. It's possible that these uninitialized bytes are eventually ending up in part of one of the `Decimal` parameters, but I'm unsure. 

There are other uninitialized reads that occur in the implementation of `std::fmt::Display` for `Decimal128`, and these occur regardless of whether or not I zero-initialize LLVM stack memory. For example:
```
let mut buf = MaybeUninit::<[c_char; decnumber_sys::DECQUAD_String]>::uninit();
let c_str = unsafe {
     if f.alternate() {
          decnumber_sys::decQuadToEngString(&self.inner, buf.as_mut_ptr() as *mut c_char);
     } else {
          decnumber_sys::decQuadToString(&self.inner, buf.as_mut_ptr() as *mut c_char);
     }
     CStr::from_ptr(buf.as_ptr() as *const c_char)
};
```
It seems like the implementation of `decQuadToString` performs a read of `buf` prior to fully initializing it. This uninitialized read occurs on line 1625 of decFloatToString
```
for (; s>=dotat; s-=4, t-=4) UBFROMUI(t, UBTOUI(s));
```
It's unclear to me if this is a problem, though. As long as `decQuadToString` initializes memory up until the null-terminator, then `CStr::from_ptr(buf.as_ptr() as *const c_char)` shouldn't behave incorrectly (I think). 
I've tried fixing this everywhere it occurs by replacing the use of `MaybeUninit<T>` with the following:
```
let mut buf = ['\0', decnumber_sys::DECQUAD_String];
```
This fixed some of the tests, but it uncovered an out-of-bounds access in `test_overloading`:
```
error: Undefined Behavior: memory access failed: alloc517857 has size 56, so pointer to 4 bytes starting at offset 54 is out-of-bounds
```
This happens on line 2247 of `decBasic.c` during the `if` condition shown below:
```
for (; *lo->msd==0 && lo->msd<lo->lsd;) lo->msd++;
if (*lo->msd==0) {
     ...
}
```
The value `0` is treated as a signed int, so the expression `*lo->msd` will read four bytes from this pointer. However, that may extend beyond the length of the allocation depending on how far the `
for` loop increments `lo->msd`. 

There's also an issue with `test_standard_notation_decimal`. It seems like the counter is reading beyond the length of the struct into the uninitialized padding at the end. Consider the following for loop in `decNumberGetBCD`:
```
uInt u=*up;                 
uInt cut=DECDPUN;              
for (; ub>=bcd; ub--) {
     *ub=(uByte)(u%10);          
     u=u/10;
     cut--;
     if (cut>0) continue;
     up++;
     u=*up; 
     cut=DECDPUN;
}
```
On the final iteration, depending on the number of digits, the expression `u=*up;` may read beyond the end of the `up`.  This error occurs as a read of uninitialized memory since there's some padding at the end of the `decNumber` struct. When I add a conditional, all tests pass, but I'm unsure if this is the best approach to fix this
```
     if (cut>0) continue;
     up++;
     if (ub - 1 >= bcd)
          u=*up; 
```
TLDR: I'll open a new PR with the changes I described that remove `MaybeUninit<T>` in favor of either `Decimal::<N>::zero()` or `['\0', decnumber_sys::DECQUAD_String]`, and t add the conditional within the for loop. With that PR and #75, all unit tests for `dec` and `decnumber-sys` pass in both Stacked Borrows and Tree Borrows except for `test_overloading`, which I was unsure of how to fix.

**Comment by benesch** [comment]
Going to close this out as the bulk of the errors have been fixed by #77. I opened #80 to track the outstanding memory unsafety in the underlying decNumber library.

Thanks again for your help here, @icmccorm!
